## ------------------------------------
##
## Script name: add metadata to single-cell object
##
## Purpose of script: Integrate samples by merging/concatenating or rPCA or CCA
##
## Author: Jiekun (Jackie) Yang
##
## Date Created: 2022-12-13
##
## Email: jkyang@mit.edu
##
##--------------------------------------

### input_file_list is a vector of file paths of r data objects
### integration_method is among "merge", "rPCA",  "CCA" and "rPCA_ref"; rPCA_ref is reference_base integration, which can improve efficiency and runtimes
### integrate_var is the variable to split the object into smaller objects for "rPCA" and "CCA"
### refs is a vector of samples to be used as references for rPCA integration
### df_name is the name of the integrated object
### feature_cutoff is the cutoff to include features detected in at least X% of cells with a default of 0.1%
### rm_mt_hsp is a logical argument indicating if reads mapped to mitochondrial genome or genes encoding heatshock proteins are removed with a default of false
### species is among "mouse" and "human"
### norm_mode is among "regular" (median normalization and log transformation) or "sctransform"
### cluster is a logical argument indicating if neighborhood identification and hierachical- and density-based clustering are performed with a default of true
### rm_prolif is a logical argument indicating if proliferating cells (the ones with MKI67 expression) are removed
### mt_cutoff allows setting a higher bound for the cells with at most X% of reads mapped to mitochondria to be included
### dims is the number of PCs used for running the UMAP, neighbor and clustering functions with a default of 50
### res is the resolution used for the clustering function with a default of 0.4
### nthreads is the number of threads used for this function
### target_folder is where you would like to store output files generated by this function

## library setup
library(Seurat)
options(future.globals.maxSize = 128000 * 1024^2)
library(future)
library(future.apply)
library(qs)
plan("multiprocess", workers = 16)
source("scripts/fancy_tsne.R")

## function
sample_integration <- function(input_file_list = NULL, integration_method = c("merge", "rPCA", "CCA", "rPCA_ref"), integrate_var = NULL, refs = NULL, target_folder = NULL, df_name = NULL, feature_cutoff = 0.001, rm_mt_hsp = FALSE, species = c("mouse", "human"), norm_mode = c("regular", "sctransform"), cluster = TRUE, nthreads = NULL, rm_prolif = TRUE, mt_cutoff = NULL, dims = 50, res = 0.4) {
  i <- 1
  for (temp.file in input_file_list) {
    temp.ext <- tools::file_ext(temp.file)
    if (temp.ext == "rds") {
      temp.data <- readRDS(temp.file)
    } else if (temp.ext == "qs") {
      temp.data <- qread(temp.file)
    }
    if (i == 1) {
      temp.all.data <- temp.data@assays$RNA@counts
    } else {
      temp.features <- intersect(rownames(temp.all.data), rownames(temp.data))
      temp.all.data <- cbind(temp.all.data[match(temp.features, rownames(temp.all.data)),], temp.data@assays$RNA@counts[match(temp.features, rownames(temp.data@assays$RNA@counts)),])
    }
    i <- i+1
  }
  temp.combined <- CreateSeuratObject(counts = temp.all.data, project = df_name, min.cells = round(feature_cutoff * ncol(temp.all.data)))
  if (species == "mouse") {
    temp.combined[["percent.mt"]] <- PercentageFeatureSet(temp.combined, pattern = "^mt-")
  } else if (species == "human") {
    temp.combined[["percent.mt"]] <- PercentageFeatureSet(temp.combined, pattern = "^MT-")
  }
  if (rm_prolif) {
    if (sum(toupper(rownames(temp.combined)) == "MKI67") > 0) {
      if (species == "mouse") {
        temp.combined <- subset(temp.combined, subset = Mki67 == 0 & percent.mt < mt_cutoff)
      } else if (species == "human") {
        temp.combined <- subset(temp.combined, subset = MKI67 == 0 & percent.mt < mt_cutoff)
      }
    } else if (sum(toupper(rownames(temp.combined)) == "PCNA") > 0) {
      if (species == "mouse") {
        temp.combined <- subset(temp.combined, subset = Pcna == 0 & percent.mt < mt_cutoff)
      } else if (species == "human") {
        temp.combined <- subset(temp.combined, subset = PCNA == 0 & percent.mt < mt_cutoff)
      }
    }
  } else {
    temp.combined <- subset(temp.combined, subset = percent.mt < mt_cutoff)
  }
  if (integration_method %in% c("rPCA", "CCA", "rPCA_ref")) {
    temp.all.data.list <- SplitObject(temp.all.data, split.by = integrate_var)
    temp.all.data.list <- lapply(X = temp.all.data.list, FUN = function(x) {
      x <- NormalizeData(x)
      x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
    })
    temp.features <- SelectIntegrationFeatures(object.list = temp.all.data.list)
    if (integration_method %in% c("rPCA", "rPCA_ref")) {
      temp.all.data.list <- lapply(X = temp.all.data.list, FUN = function(x) {
      x <- ScaleData(x, features = temp.features, verbose = FALSE)
      x <- RunPCA(x, features = temp.features, verbose = FALSE)
      })
    }
    if (integration_method %in% c("rPCA", "CCA")) {
      temp.anchors <- FindIntegrationAnchors(object.list = temp.all.data.list, anchor.features = temp.features, reduction = tolower(integration_method))
    } else {
      temp.anchors <- FindIntegrationAnchors(object.list = temp.all.data.list, reference = refs, reduction = "rPCA")
    }
    temp.combined <- IntegrateData(anchorset = temp.anchors)
    DefaultAssay(temp.combined) <- "integrated"
  }
  if (ncol(temp.combined) > 100000) {
    temp.combined <- fancy_tsne(temp.combined, TRUE, rm_mt_hsp, species, norm_mode, cluster, nthreads, dims = dims, res = res)
  } else {
    temp.combined <- fancy_tsne(temp.combined, FALSE, rm_mt_hsp, species, norm_mode, cluster, nthreads, dims = dims, res = res)
  }
  # visually check expression of a few very specifically expressed marker genes
  png(paste0(target_folder, "/plots/", df_name, "_QC_check.png"), width = 2500, height = 500)
  print(FeaturePlot(temp.combined, features = c("Ptprc", "Hba-a1", "nCount_RNA", "nFeature_RNA", "percent.mt"), ncol = 5))
  dev.off()
  qsave(temp.combined, paste0(target_folder, "/rdata/", df_name, "_combined.qs"))
  return(temp.combined)
}
